### ISSUE TITLE
backward for tensor.min() and tensor.min(dim=0) behaves differently

### ISSUE THREAD

**Issue by gkioxari** [pytorch/pytorch#35699]
## Backward for `tensor.min` behaves differently if `dim` is set 

I noticed that the gradient of the `tensor.min()` function gives a different output when `dim` is set. Namely, when `dim` is set, the gradient will only correspond to one of the entries with the lowest values. If `dim` is not set, then there will be gradient for all entries which have the lowest value

## Code sample to reproduce

```
import torch
a = torch.tensor([0.1, 0.3, 0.1], dtype=torch.float32, requires_grad = True)
a_cp  = torch.tensor([0.1, 0.3, 0.1], dtype=torch.float32, requires_grad = True)
b = a.min()
b.backward()
a.grad  # Output is tensor([1., 0., 1.])
c, d = a_cp.min(dim=0)
c.backward()
a_cp.grad  # Output is tensor([0., 0., 1.])
```

## Problematic for vectorized implementations

This behavior is not expected and it leads to discrepancies in the gradients between vectorized and non implementations. Setting `dim` in `torch.min()` should not change the gradients. 

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen

**Comment by ailzhang** [comment]
I agree that we should fix this discrepancy in behaviors. 
Marking triage review so that we can agree on whether fixing `torch.min()` or `torch.min(dim=0)`.

**Comment by jcjohnson** [comment]
The `torch.min(dim=0)` behavior is more correct, since it's a subgradient of the function at this nondifferentiable point.

See for example:
https://see.stanford.edu/materials/lsocoee364b/01-subgradients_notes.pdf

**Comment by gchanan** [comment]
I believe the `torch.min(dim=0)` is also the behavior of related functions like max_pool.

**Comment by albanD** [comment]
Yes only one of the minimum value should get gradient.
So we should change torch.min().

**Comment by nikitaved** [comment]
`min.dim` uses `scatter_` in the backward, hence it is not deterministic. What would be preferred behavior in here? Shall it mirror the behavior of argmin/argmax?

**Comment by albanD** [comment]
The problem is not with determinism. It is with the fact that we set multiple values to the gradient and not a single one.

I think this is more general than just min. `select_equals_backward` is not the right thing to use for min/max. Only one of the entries should get gradients, not all of them as we do: `masked_fill_(input == value, grad)`. Not sure how to implement that efficiently though...

**Comment by nikitaved** [comment]
I am totally aware of the problem. The question is which coordinate gets the gradient? Shall it be deterministic? The current implementation for min uses 'scatter' so, in backward, it is unknown which value gets the gradient.

**Comment by albanD** [comment]
The general rule is that we try to keep it deterministic. But if it is more efficient not to be, then it is fine. You can see we mention it in the doc [here](https://pytorch.org/docs/stable/notes/randomness.html#pytorch).
In this particular case, correctness is more important than determinism.

But even in a non-deterministic way, what do you think is an efficient implementation of this backward?

**Comment by nikitaved** [comment]
I think it makes sense to follow the approach from argmin/argmax. Two custom kernels for CPU and GPU. CPU in a simple loop (maybe two loops, but all of the kernalized), and for GPU copying the approach from argmax/argmin. Does it make any sense?

**Comment by albanD** [comment]
Ho right, writing custom kernels will definitely work 👍 
If you write a custom kernel, you will need to write a backward for that backward as well.
I was more thinking of a generic backward like the one we have right now.

But if there is no efficient solution with the primitives we already have I guess we can do this.
Another thing I was thinking is to make the `min()` op return the index of one minimum value that we use for the autograd (but do not return to the user, like we do for maxpool) and the backward is then simply writing the grad value at that index.

**Comment by nikitaved** [comment]
> Another thing I was thinking is to make the `min()` op return the index of one minimum value that we use for the autograd (but do not return to the user, like we do for maxpool) and the backward is then simply writing the grad value at that index.

Yeah, exactly what I was thinking about to reuse `min.dim`. It will not be a problem to hide the indices, will it? But then I am not sure that the indices that 'min.dim' returns are deterministic :)
So, determinism is the priority here, right?

**Comment by nikitaved** [comment]
It could be that `scatter` will become determinism-friendly at some point... So maybe reusing `min.dim` is not a bad idea...

**Comment by albanD** [comment]
> So, determinism is the priority here, right?

No I don't think so.
1) Correctness
2) Speed/memory
3) Determinism

> to reuse min.dim

Maybe the simplest way is this yes.
Completely remove min() from the derivatives.yaml and implement it with min.dim. You just need to make sure that the implementation works in all cases that min() support (non-contiguous inputs, sparse, etc).

**Comment by nikitaved** [comment]
@albanD , I am a bit confused with all these autogenerated `min` variants. Could you please shine some light here? I see that there is a `min` variant with 3 dispatches. Now, there are other `min` variants, including the ones with `dim` specified. I guess these are generated, right? Am I correct that the original implementation `min` with `dim` specified traces back to `_min`?
What's the difference between `min` and `_min`. It seems like they are using different dispatches, but why? Is it before of performance?

**Comment by nikitaved** [comment]
I think at this point it could be easier to modify the backward method where first the forward is done with the indices returned, and then the backward similar to `min.dim`. Sounds like a hack, but at least it is possible to reuse the already implemented functionality. Ah, forgot, it will require a new dispatch for quantized...

**Comment by albanD** [comment]
From the autograd point of view (in derivatives.yaml), we have 3 versions:
- min() that does the full reduction
- min(dim=) that does reduction on a given set of dimensions
- min(other=) that does element-wise minimum between Tensors.

Only the first one is problematic. (Note that the version of max() that does the full reduction uses the same backward and so has the same issue).


From the point of view of aten (native_functions.yaml)
- min() that does the full reduction (with a special quantized dispatch as you mentionned).
- min.dim (with the named versions and the versions where the user provide the output Tensor and the version with only the values and no indices)
- min.other (and the version where the user provide the output Tensor as min.out)
- _min which are the TH implementations if the dim version (and the out= variant)

The simplest fix I think is:
```
Tensor select_equals_backward(Tensor grad, const Tensor & input, const Tensor & value) {
  auto grad_input = at::zeros_like(input, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
  // Below is pseudo-code, might not run as is
  auto first_nonzero = (input == value).nonzero().select(0, 0);
  auto tmp = grad_input;
  for (auto dim=0; dim < grad_input.sizes().size(); ++dim) {
    tmp = tmp.select(0, first_nonzero[dim].item()); // Not sure what is the cpp equivalent of item
  }
  tmp.copy_(grad);
  return grad_input;
}
```

**Comment by nikitaved** [comment]
Thanks for the help, now the structure makes much more sense.
But am I correct that this function assumes batched input? Then the loop has to be over the batch dimension, right?

**Comment by albanD** [comment]
No the full reduction goes from ND to a scalar. So the backward takes a scalar and returns an ND Tensor.
There is no notion of batch for this function.

In python you would do `grad_input[*first_nonzero] = grad` but I don't think you can do that in cpp. So I just index each dimension one by one in a for-loop :D

**Comment by nikitaved** [comment]
@albanD , looks like `select` creates a new tensor. So, my plan is to create an array of 0-dim Tensors from `nonzero` and then use `index_put`. Would that work?

**Comment by albanD** [comment]
`.select()` creates a new Tensor but it shares storage and so is always a view of the original one.
So the code above should work.
Not sure how you would do with index_put.

**Comment by nikitaved** [comment]
OK, indeed, `select` does the job. Thank you!

**Comment by albanD** [comment]
PR has been merged. This should be all fine now!
